tls/s2n_server_hello.c

```diff
@@ -76,14 +87,31 @@ int s2n_server_hello_recv(struct s2n_connection *conn)
         if (s2n_is_hello_retry_req(conn)) {
             GUARD(s2n_server_hello_retry_recv(conn));
         }
+
         /* Check echoed session ID matches */
         S2N_ERROR_IF(session_id_len != conn->session_id_len || memcmp(session_id, conn->session_id, session_id_len), S2N_ERR_BAD_MESSAGE);
         conn->actual_protocol_version = conn->server_protocol_version;
         GUARD(s2n_set_cipher_as_client(conn, cipher_suite_wire));
     } else {
         uint8_t actual_protocol_version;
-
         conn->server_protocol_version = (uint8_t)(protocol_version[0] * 10) + protocol_version[1];
+
+        /* Detect downgrade attacks according to RFC 8446 section 4.1.3 */
+        uint8_t *downgrade_protection = &conn->secure.server_random[S2N_TLS_RANDOM_DATA_LEN - S2N_DOWNGRADE_PROTECTION_SIZE];
+        if (s2n_is_tls13_enabled() && conn->server_protocol_version == S2N_TLS12) {
+            /* If the server supports TLS1.3, the random data will be set according to RFC sect 4.1.3 */
+            if (memcmp(downgrade_protection, tls12_downgrade_protection_bytes, S2N_DOWNGRADE_PROTECTION_SIZE) == 0) {
+                GUARD(s2n_queue_reader_illegal_parameter_alert(conn));
+                S2N_ERROR(S2N_ERR_SERVER_PROTOCOL_VERSION);
+            }
+        } else if (conn->server_protocol_version < S2N_TLS12) {
```

Comment:
The RFC says "TLS 1.2 servers SHOULD" set these values. Can we safely assume all TLS 1.2 servers will?

Reply:
We can not. They will all send their random value, but they aren't strictly required to indicate they support a later version than they negotiate.